File: apps/api/src/hoba_api/wheel/spin_math.py
"""Pure wheel math — Python mirror of `apps/webapp/src/features/wheel/spinMath.ts`.

JS bitwise semantics emulated by masking everything to unsigned 32-bit:
- `Math.imul(a, b)` ≡ `(a * b) & 0xFFFFFFFF` (low 32 bits identical
  whether operands are interpreted signed or unsigned — two's
  complement multiplication preserves low bits).
- `a >>> n` ≡ `a >> n` once `a` is masked to unsigned 32-bit.
- `(a + b) | 0` and `a ^= b` both coerce to 32-bit ⇒ mask after.
"""

from __future__ import annotations

from collections.abc import Callable
from dataclasses import dataclass

UINT32_MASK = 0xFFFF_FFFF
UINT32_RANGE = 0x1_0000_0000  # 2**32

MIN_DURATION_MS = 5_500
MAX_DURATION_MS = 8_500
MIN_FULL_ROTATIONS = 5
MAX_FULL_ROTATIONS = 8
SECTOR_JITTER_FRACTION = 0.7


def mulberry32(seed: int) -> Callable[[], float]:
    """Deterministic PRNG yielding uniform floats in [0, 1)."""
    state = [seed & UINT32_MASK]

    def next_() -> float:
        state[0] = (state[0] + 0x6D_2B_79_F5) & UINT32_MASK
        t = state[0]
        t = ((t ^ (t >> 15)) * (t | 1)) & UINT32_MASK
        inner = ((t ^ (t >> 7)) * (t | 61)) & UINT32_MASK
        t = (t ^ ((t + inner) & UINT32_MASK)) & UINT32_MASK
        return (t ^ (t >> 14)) / UINT32_RANGE

    return next_


def pick_segment_index(
    segment_count: int,
    rng: Callable[[], float],
    weights: list[float] | None = None,
) -> int:
    """Weighted-or-uniform random index in `[0, segment_count)`."""
    if segment_count <= 0:
        raise ValueError("segment_count must be > 0")
    if not weights:
        return min(int(rng() * segment_count), segment_count - 1)
    if len(weights) != segment_count:
        raise ValueError(
            f"weights length ({len(weights)}) ≠ segment_count ({segment_count})",
        )
    total = sum(max(0.0, w) for w in weights)
    if total <= 0:
        return min(int(rng() * segment_count), segment_count - 1)
    r = rng() * total
    for i, w in enumerate(weights):
        r -= max(0.0, w)
        if r <= 0:
            return i
    return segment_count - 1


@dataclass(frozen=True, slots=True)
class SpinResult:
    result_segment_index: int
    final_angle_deg: float
    duration_ms: int
    seed: int


def compute_spin(
    *,
    segment_count: int,
    seed: int,
    starting_angle_deg: float = 0.0,
    winning_index: int | None = None,
    weights: list[float] | None = None,
) -> SpinResult:
    """Resolve a spin. Returns the absolute target rotation.

    Always travels at least `MIN_FULL_ROTATIONS × 360°` clockwise from
    `starting_angle_deg` — never less, never backwards.
    """
    if not (2 <= segment_count <= 12):
        raise ValueError(
            f"segment_count must be in [2, 12], got {segment_count}",
        )

    rng = mulberry32(seed)

    if winning_index is None:
        winning_index = pick_segment_index(segment_count, rng, weights)

    sector_deg = 360 / segment_count
    sector_jitter = (rng() - 0.5) * sector_deg * SECTOR_JITTER_FRACTION
    rotations = MIN_FULL_ROTATIONS + int(
        rng() * (MAX_FULL_ROTATIONS - MIN_FULL_ROTATIONS + 1),
    )

    sector_center_offset = winning_index * sector_deg + sector_deg / 2
    target_mod = ((-sector_center_offset + sector_jitter) % 360 + 360) % 360
    start_mod = (starting_angle_deg % 360 + 360) % 360
    base_delta = (target_mod - start_mod + 360) % 360
    total_delta = base_delta + rotations * 360
    final_angle = starting_angle_deg + total_delta

    duration_ms = MIN_DURATION_MS + int(
        rng() * (MAX_DURATION_MS - MIN_DURATION_MS),
    )

    return SpinResult(
        result_segment_index=winning_index,
        final_angle_deg=final_angle,
        duration_ms=duration_ms,
        seed=seed,
    )


def segment_under_pointer(rotation_deg: float, segment_count: int) -> int:
    """Inverse: which segment is currently under the top pointer."""
    sector_deg = 360 / segment_count
    angle = (-rotation_deg % 360 + 360) % 360
    return int(angle // sector_deg) % segment_count

File: apps/webapp/src/features/wheel/spinMath.ts
/**
 * Pure math for the wheel spin.
 *
 * Convention: segment 0 starts at the 12 o'clock pointer and segments
 * are laid out clockwise around the wheel. Each segment occupies
 * `sectorDeg = 360 / segmentCount`. To land segment K under the
 * (fixed) top pointer, we rotate the wheel clockwise by an angle whose
 * `mod 360` brings the midpoint of K's sector to 0°. We add several
 * full revolutions on top for dramatic effect.
 */

import { mulberry32 } from "./seededRandom";
import type { SpinResult } from "./types";

const MIN_DURATION_MS = 5_500;
const MAX_DURATION_MS = 8_500;
const MIN_FULL_ROTATIONS = 5;
const MAX_FULL_ROTATIONS = 8;
/** Random offset within the winning sector — keeps landings off dead-center. */
const SECTOR_JITTER_FRACTION = 0.7;

/** Weighted-or-uniform random index in `[0, segmentCount)`. */
export function pickSegmentIndex(
  segmentCount: number,
  rng: () => number,
  weights?: number[],
): number {
  if (segmentCount <= 0) {
    throw new Error("pickSegmentIndex: segmentCount must be > 0");
  }
  if (weights === undefined || weights.length === 0) {
    return Math.min(Math.floor(rng() * segmentCount), segmentCount - 1);
  }
  if (weights.length !== segmentCount) {
    throw new Error(
      `pickSegmentIndex: weights length (${weights.length}) ≠ segmentCount (${segmentCount})`,
    );
  }
  const total = weights.reduce((sum, w) => sum + Math.max(0, w), 0);
  if (total <= 0) {
    return Math.min(Math.floor(rng() * segmentCount), segmentCount - 1);
  }
  let r = rng() * total;
  for (let i = 0; i < segmentCount; i++) {
    r -= Math.max(0, weights[i] ?? 0);
    if (r <= 0) return i;
  }
  return segmentCount - 1;
}

export interface ComputeSpinInput {
  segmentCount: number;
  seed: number;
  /**
   * Absolute current rotation of the wheel (the MotionValue's value).
   * Defaults to 0 for first spins. **Subsequent spins must pass the
   * wheel's current rotation** — otherwise the new animation may travel
   * a tiny arc forward or even rotate backwards (the wheel reverses
   * after first spin bug).
   */
  startingAngleDeg?: number;
  /** When provided, RNG only contributes to angle jitter + rotations + duration. */
  winningIndex?: number;
  /** For Rigged Mode in later phases. Length must equal `segmentCount`. */
  weights?: number[];
}

/**
 * Resolve a spin given a `{segmentCount, seed}` pair.
 *
 * Returns an **absolute** target rotation (`finalAngleDeg`) such that
 * animating the wheel to it lands `winningIndex` under the top pointer.
 * Always travels at least `MIN_FULL_ROTATIONS × 360°` clockwise from
 * `startingAngleDeg` — never less, never backwards.
 */
export function computeSpin(input: ComputeSpinInput): SpinResult {
  const { segmentCount, seed, weights, startingAngleDeg = 0 } = input;
  if (segmentCount < 2 || segmentCount > 12) {
    throw new Error(
      `computeSpin: segmentCount must be in [2, 12], got ${segmentCount}`,
    );
  }

  const rng = mulberry32(seed);

  const winningIndex =
    input.winningIndex ?? pickSegmentIndex(segmentCount, rng, weights);

  const sectorDeg = 360 / segmentCount;
  const sectorJitter = (rng() - 0.5) * sectorDeg * SECTOR_JITTER_FRACTION;
  const rotations =
    MIN_FULL_ROTATIONS +
    Math.floor(rng() * (MAX_FULL_ROTATIONS - MIN_FULL_ROTATIONS + 1));

  // Rotation (mod 360) that places `winningIndex` under the top pointer:
  // pointer reads K iff `((-R) mod 360)` is in K's sector — centered at
  // `K * sectorDeg + sectorDeg/2`. Jitter shifts within the sector.
  const sectorCenterOffset = winningIndex * sectorDeg + sectorDeg / 2;
  const targetMod =
    (((-sectorCenterOffset + sectorJitter) % 360) + 360) % 360;
  const startMod = ((startingAngleDeg % 360) + 360) % 360;
  const baseDelta = (targetMod - startMod + 360) % 360;
  const totalDelta = baseDelta + rotations * 360;
  const finalAngleDeg = startingAngleDeg + totalDelta;

  const durationMs =
    MIN_DURATION_MS + Math.floor(rng() * (MAX_DURATION_MS - MIN_DURATION_MS));

  return { resultSegmentIndex: winningIndex, finalAngleDeg, durationMs, seed };
}

/**
 * Inverse: given a rotation (the live MotionValue), report which
 * segment is currently under the top pointer. Used by the rAF tick
 * detector during the decelerate phase.
 */
export function segmentUnderPointer(
  rotationDeg: number,
  segmentCount: number,
): number {
  const sectorDeg = 360 / segmentCount;
  // The pointer sits at 0°; after wheel rotation `r`, the angle at the
  // pointer relative to the (now-rotated) wheel is `-r`. Normalize to
  // [0, 360) and find which sector contains it.
  const angle = ((-rotationDeg) % 360 + 360) % 360;
  return Math.floor(angle / sectorDeg) % segmentCount;
}
